Implement uring in `AsyncWrite for fs::File`

RedKinda opened the pull request:
## Motivation

Currently IO operations for `fs::File` is implemented using `spawn_blocking` to perform IO, which is not great for an async runtime. This PR implements IO operations using `io_uring`. The `fs::write` function already uses `io_uring` for async operations, and this PR extends the uring functionality into `impl AsyncWrite for fs::File`.

For full discussion see here https://github.com/tokio-rs/tokio/discussions/7684 and subsequently on discord https://discord.com/channels/500028886025895936/810724255046172692/1427735341003051079

## Solution

The `fs::File` struct already implements an internal buffer when writing using `spawn_blocking`, which is similar to what `io_uring` requires. Therefore in `poll_write`, we can check if `io_uring` is initialized and supported, and if not fall back to the `spawn_blocking` implementation, without requiring any API changes.

This PR only implements `AsyncWrite`, and once this PR is accepted, I am happy to also open a PR for `AsyncRead for File` using the same structure.

RedKinda committed:
implement uring AsyncWrite for File

RedKinda committed:
add basic uring File write tests

RedKinda committed:
formatting

martin-g commented on tokio/src/fs/file.rs:
```diff
@@ -753,20 +808,75 @@ impl AsyncWrite for File {
+                                    let handle = BoxedOp(Box::pin(async move {
+                                        let (r, buf, _fd) = op.await;
+                                        match r {
+                                            Ok(_n) => (Operation::Write(Ok(())), buf),
```
I think there is a chance that the buffer is not fully drained here.
`Write(Ok())` should be returned only if `buf.is_empty()`.

I think you need to wrap the write in a loop and handle it with something like:
```rust
Ok(0) => break (Operation::Write(Err(io::ErrorKind::WriteZero.into())), buf),
Ok(_) if buf.is_empty() => break (Operation::Write(Ok(())), buf),
Ok(_) => continue, // more to write
Err(e) => break (Operation::Write(Err(e)), buf),
```

martin-g commented on tokio/src/fs/file.rs:
```diff
@@ -824,20 +934,73 @@ impl AsyncWrite for File {
+                                    let handle = BoxedOp(Box::pin(async move {
+                                        let (r, buf, _fd) = op.await;
+                                        match r {
+                                            Ok(_n) => (Operation::Write(Ok(())), buf),
```
Same as above - make sure the buf is empty before saying Ok().

RedKinda committed:
remove sync bound from BoxedOp

RedKinda committed:
make sure to write the full buffer in uring file writes

RedKinda commented on tokio/src/fs/file.rs:
```diff
@@ -107,7 +107,61 @@ struct Inner {
+}
+
+cfg_io_uring! {
+    struct BoxedOp<T>(Pin<Box<dyn Future<Output = T> + Send + Sync + 'static>>);
```
Looks like it does need Sync, at least its required by a test https://github.com/RedKinda/tokio/blob/uring_file/tokio/tests/async_send_sync.rs#L651 - https://github.com/tokio-rs/tokio/actions/runs/18909744941/job/53976898682?pr=7713#step:6:248

RedKinda committed:
Revert "remove sync bound from BoxedOp"

This reverts commit 863a69c8599e390dd273115693e3349c2432bdb0.

RedKinda commented:
This should be ready for merging i think, is there anything else blocking it?

Darksonn commented on tokio/src/fs/file.rs:
```diff
@@ -753,20 +808,90 @@ impl AsyncWrite for File {
+                            if handle.inner.driver().io().check_and_init()? {
+                                task_join_handle = {
+                                    use crate::{io::uring::utils::ArcFd, runtime::driver::op::Op};
+
```
This is way too many levels of indentation. We need to pull some logic into a separate function.

RedKinda commented on tokio/src/fs/file.rs:
```diff
@@ -107,7 +107,61 @@ struct Inner {
+        feature = "fs",
+        target_os = "linux"
+    ))]
+    Async(BoxedOp<T>),
```
The `Blocking(JoinHandle)` is `crate::blocking::JoinHandle` not `tokio::runtime::task::join::JoinHandle` so some enum would be required anyway, and at that point might as well just have this kind of a boxed future

Darksonn replied:
That is a re-export of the same thing.

https://github.com/tokio-rs/tokio/blob/4714ca168d6bd97193625657b0381e9b65a9ceff/tokio/src/blocking.rs#L9

RedKinda replied:
Ah got it, I'll change it.

Out of curiosity, general rule is that for a future to keep making progress, user needs to keep polling it. I feel like not having to do that specifically for `fs::File` is just an implementation detail, and users shouldn't rely on that always being the case. So i guess my question is how much is it worth preserving old behavior like this when implementing new features, if it comes at the cost of (small) performance overhead? [relevant XKCD](https://xkcd.com/1172/)

RedKinda replied:
Had to adjust mocks to accommodate non-blocking spawning https://github.com/tokio-rs/tokio/actions/runs/19464918802/job/55697554687 https://github.com/tokio-rs/tokio/pull/7713/commits/b91836c93f3e0a1d669848d3e807536a07aa92e4 - can you doublecheck im doing it right?

RedKinda committed:
move part of poll_write to its own function

martin-g commented on tokio/src/fs/write.rs:
```diff
@@ -38,10 +45,15 @@ pub async fn write(path: impl AsRef<Path>, contents: impl AsRef<[u8]>) -> io::Re
         let driver_handle = handle.inner.driver().io();
         if driver_handle.check_and_init()? {
-            return write_uring(path, contents).await;
+            use crate::io::blocking;
```
Is this import needed here ?
It is already imported at line 8

RedKinda committed:
implement review comment suggestions

RedKinda committed:
add non-blocking spawn to mocks

RedKinda committed:
fix type issue in mock spawn

RedKinda committed:
fix spawn signature 2.0

RedKinda committed:
appease clippy

RedKinda committed:
Merge branch 'master' into uring_file

RedKinda committed:
Merge branch 'master' into uring_file

RedKinda committed:
Merge branch 'master' into uring_file

Darksonn commented on tokio/src/io/uring/write.rs:
```diff
@@ -29,27 +41,20 @@ impl Cancellable for Write {
-        file_offset: u64,
-    ) -> io::Result<Self> {
+    /// Issue a write at `file_offset` from the provided `buf`. To use current file cursor, set `file_offset` to `-1` or `u64::MAX`.
+    pub(crate) fn write_at(fd: ArcFd, buf: blocking::Buf, file_offset: u64) -> Self {
```
This seems to be a feature that relies on a flag `IORING_FEAT_RW_CUR_POS` being present. We should probably make sure to check this feature somehow.

RedKinda committed:
Merge remote-tracking branch 'origin/master' into uring_file

RedKinda committed:
reuse poll_write_inner, fix lint

RedKinda committed:
sync uring writes with shape of uring reads

RedKinda committed:
remove unused tests

RedKinda committed:
make Write uring op generic over fd type

RedKinda commented:
sorry for the delay, I got busy with life and forgot I was working on this. I have updated the PR to be consistent with #7907 
- i removed the mock tests since they arent included in the AsyncRead PR, happy to add some more testing if we know what that should look like
- i made Write generic over the Fd type. Read op is also generic over buf type, didn't feel like it was needed for Writes though so I didn't include it
